python/whylogs/extras/nlp_metric.py

```diff
@@ -0,0 +1,289 @@
+            S=VectorComponent(np.zeros(1)),
+        )
+
+
+@dataclass(frozen=True)
+class UpdatableSvdMetric(SvdMetric):
+    """
+    updating SVD metric
+    """
+
+    @property
+    def namespace(self) -> str:
+        return "updatable_svd"
+
+    def _resketch(self, k: int, decay: float, U1: np.ndarray, S1: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+        U0, S0 = self.U.value, self.S.value
+        if U0.shape == (0, 0):
+            U0 = np.zeros((U1.shape[0], k))
+            S0 = np.zeros(k)
+        if U0.shape[0] < U1.shape[0]:
+            U0 = np.pad(U0, ((0, U1.shape[0] - U0.shape[0]), (0, 0)), "constant")
+        assert U0.shape[0] == U1.shape[0]
+
+        Q, R = np.linalg.qr(np.concatenate((decay * U0 * S0, U1 * S1), axis=1))
+        UR, S, VRT = np.linalg.svd(R)
+        U = np.dot(Q, UR)
+        return U, S
+
+    def merge(self, other: "SvdMetric") -> "UpdatableSvdMetric":
+        # other can be updatable or not
+        new_U, new_S = self._resketch(self.k.value, self.decay.value, other.U.value, other.S.value)
+        return UpdatableSvdMetric(self.k, self.decay, VectorComponent(new_U), VectorComponent(new_S))
+
+    def columnar_update(self, data: PreprocessedColumn) -> OperationResult:
+        k = self.k.value
+        decay = self.decay.value
+        vectors_processed = 0
+        for vector in data.list.objs:
+            if (not isinstance(vector, np.ndarray)) or vector.shape[0] < 2:
+                continue
+
+            # TODO: batch this
+            vectors_processed += 1
+            U1, S1, _ = np.linalg.svd(vector.reshape((vector.shape[0], 1)), False, True, False)
+            new_U, new_S = self._resketch(k, decay, U1, S1)
+            self.U.set(new_U)
+            self.S.set(new_S)
+
+        return OperationResult.ok(vectors_processed)
+
+    @classmethod
+    def zero(cls, config: SvdMetricConfig) -> "UpdatableSvdMetric":
+        if config is None or not isinstance(config, SvdMetricConfig):
+            raise ValueError("UpdatableSvdMetric.zero() requires SvdMetricConfig argument")
+
+        return UpdatableSvdMetric(
+            k=IntegralComponent(config.k),  # not usable with k = 0
+            decay=FractionalComponent(config.decay),
+            U=VectorComponent(np.zeros((1, 1))),
+            S=VectorComponent(np.zeros(1)),
+        )
+
+
+@dataclass(frozen=True)
+class NlpConfig(MetricConfig):
+    """
+    If you pass in an UpdatableSvdMetric, the SVD will be updated along with the
+    NlpMetric's residual distribution. A non-updatable SvdMetric will update the
+    residual distribution, but it will not update the SVD as new term vectors are
+    processed.
+
+    Note that the [Updatable]SvdMetric is not [de]serialized with the NlpMetric.
+    You'll have to manage that yourself.
+    """
+
+    # The default will not allow updates or residual computation
+    svd: SvdMetric = SvdMetric.zero(SvdMetricConfig())
+
+
+@dataclass
+class NlpMetric(CompoundMetric):
```

Comment:
An example notebook that calls this from why.log* would be useful context, I think you basically have that in your example.py modulo the top level api integration and callstack. Also some tests would be useful to better see how this works. Is it easy to configure nlp metrics to be enabled? and how do you want callers to log documents?